File: python/1105. Filling Bookcase Shelves.py
class Solution(object):
    def minHeightShelves(self, books, shelf_width):
        """
        :type books: List[List[int]]
        :type shelf_width: int
        :rtype: int
        """
        n = len(books)
        dp = [float('inf')] * (n + 1)
        dp[0] = 0
        for i in range(1, n+1):
            w = h = 0
            for j in range(i, 0, -1):
                w += books[j-1][0]
                if w > shelf_width:
                    break
                h = max(h, books[j-1][1])
                dp[i] = min(dp[i], dp[j-1] + h)
        return dp[n]

File: go/src/leetcode/1105/1105. Filling Bookcase Shelves.go
package main

import "math"

/*
*
https://leetcode.cn/problems/filling-bookcase-shelves/description/
*/
func minHeightShelves(books [][]int, shelfWidth int) int {
	n := len(books)
	f := make([]int, n+1)
	for i := range f {
		f[i] = math.MaxInt32
	}
	f[0] = 0
	for i := 1; i <= n; i++ {
		var w, h int
		for j := i - 1; j >= 0; j-- {
			w += books[j][0]
			if w > shelfWidth {
				break
			}
			h = max(h, books[j][1])
			f[i] = min(f[i], f[j]+h)
		}
	}
	return f[n]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
